### PR TITLE
[ENH] Quantile metric design proposal

### PR DESCRIPTION
**Pull request by eenticott-shell**
#### Reference Issues/PRs
Addresses #1763 

#### What does this implement/fix? Explain your changes.
A prototype of a new metric class designed for probabilistic or quantile inputs. The specific implementation is for the pinball loss.

#### What should a reviewer concentrate their feedback on?

The overall structure of the class, public and private method specification, defaults and handling of input types. Some of this remains to be properly implemented. 

##### For all contributions
- [ ] I've added myself to the [list of contributors](https://github.com/alan-turing-institute/sktime/blob/main/.all-contributorsrc).
- [ ] Optionally, I've updated sktime's [CODEOWNERS](https://github.com/alan-turing-institute/sktime/blob/main/CODEOWNERS) to receive notifications about future changes to these files.
- [ ] I've added unit tests and made sure they pass locally.

### CODE REVIEW

**Review by fkiraly — CHANGES_REQUESTED** [review]
I assume this is still work in progress?

In terms of design ideas, looks good for me, but this needs to be cleaned up.

* pinball loss won't work for `multioutput`, will it?
* `I don't think `if y_true >= y_pred` runs in either of the places where you compute the loss. You want to vectorize the computation.
* should `_evaluate` of the pinball loss call a vectorized function `pinball_loss`?
* the base class should do input checks and/or conversions
* you'll need to iterate over multiple variables
* please add docstrings
* please add tests

**Comment by eenticott-shell** [comment]
Closing as addressed by #1981

File: sktime/performance_metrics/forecasting/probabilistic/_classes.py [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+import numpy as np
+
+# TODO: add formal tests
+import pandas as pd
+from sklearn.metrics._regression import _check_reg_targets
+
+from sktime.performance_metrics.forecasting._classes import _BaseForecastingErrorMetric
+
+
+def _check_y_pred_type(metric_object, y_pred):
```
Comment (fkiraly) [inline sktime/performance_metrics/forecasting/probabilistic/_classes.py:14]: good idea - I guess this is where we would put checks and the conversions?

There's currently one conversion - `_convert_pred_interval_to_quantiles` in `forecasting.base._base`, and one check, `_check_predict_quantiles`. There is no check for the new "predict interval" format.

File: sktime/performance_metrics/forecasting/probabilistic/_classes.py [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+    """Base class for probabilistic forecasting error metrics in sktime.
+
+    Extends sktime's BaseMetric to the forecasting interface. Forecasting error
+    metrics measure the error (loss) between forecasts and true values. Lower
+    values are better.
+    """
+
+    _tags = {"scitype:y_pred": "quantiles"}
+
+    def __init__(self, func, name=None, multioutput="uniform_average"):
```
Comment (fkiraly) [inline sktime/performance_metrics/forecasting/probabilistic/_classes.py:32]: what is the "func" being used for, is that really necessary?

File: sktime/performance_metrics/forecasting/probabilistic/_classes.py [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+    ----------
+    y_true : np.array of shape (fh,) or (fh, n_outputs) \
+             where fh is the forecasting horizon
+        Ground truth (correct) target values.
+
+    y_pred : np.array of shape (fh,) or (fh, n_outputs) \
+             where fh is the forecasting horizon
+        Forecasted alpha quantiles
+
+    alpha : The significance
```
Comment (fkiraly) [inline sktime/performance_metrics/forecasting/probabilistic/_classes.py:154]: not a significance here, but the fraction at which the quantile is computed (not sure what the right term is, it's not the quantile itself)

Comment (fkiraly) [reply sktime/performance_metrics/forecasting/probabilistic/_classes.py:154]: I think it would make sense to allow vectorization of alpha? Then do `np.array` casting?

File: sktime/performance_metrics/forecasting/probabilistic/_classes.py [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+    """
```
Comment (fkiraly) [inline sktime/performance_metrics/forecasting/probabilistic/_classes.py:155]: docstring is missing a `Returns` section

**Review by fkiraly — COMMENTED** [review]
I think this is good, could be turned into a "non-draft" PR.

I left some comments, mainly about docstrings, vectorization, and parameters.

File: sktime/performance_metrics/forecasting/probabilistic/_classes.py (hunk 2 above)
Comment (eenticott-shell) [reply sktime/performance_metrics/forecasting/probabilistic/_classes.py:32]: The `func` contains the underlying logic implementation for the metric. It is necessary as I am inheriting from the previous metric class which included it. It can be useful if you need two different metric classes that use the same function in different ways.

File: sktime/performance_metrics/forecasting/probabilistic/_classes.py (hunk 1 above)
Comment (eenticott-shell) [reply sktime/performance_metrics/forecasting/probabilistic/_classes.py:14]: Yes this is mostly a placeholder for now. I'll likely add to this in the next PR (adding interval metrics)

File: sktime/performance_metrics/forecasting/probabilistic/_classes.py (hunk 4 above)
Comment (eenticott-shell) [reply sktime/performance_metrics/forecasting/probabilistic/_classes.py:198]: I have added this.

File: sktime/performance_metrics/forecasting/probabilistic/_classes.py (hunk 2 above)
Comment (fkiraly) [reply sktime/performance_metrics/forecasting/probabilistic/_classes.py:34]: hm, ok - is there a way to make it optional?

I'm thinking about extenders, who will probably fund it very confusing tha `pinball_loss` is passed in two places, and `func` isn't really used in the child, `PinballLoss`.
For instance, why can `_evaluate` not take the place of `func`? Or the other way round?

File: sktime/performance_metrics/forecasting/probabilistic/_classes.py [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+    ----------
+    alpha : The significance level required for the pinball loss
+
+    multioutput : string "uniform_average" or "raw_values" determines how multioutput
+    results will be treated
+    """
+
+    def __init__(self, alphas, multioutput="uniform_average"):
+        name = "PinballLoss"
+        # _tags = {"scitype:pred": "quantile"}
```
Comment (fkiraly) [inline sktime/performance_metrics/forecasting/probabilistic/_classes.py:102]: this should be without comment, and directly in the class, no?

File: sktime/performance_metrics/forecasting/probabilistic/_classes.py [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+        func = pinball_loss
+        self.alphas = self._check_alpha(alphas)
+        super().__init__(func=func, name=name, multioutput=multioutput)
+
+    def _check_alpha(self, alpha):
```
Comment (fkiraly) [inline sktime/performance_metrics/forecasting/probabilistic/_classes.py:107]: would it make sense to put this in a quantile mixin, or in the base class? Other quantile losses will need this too?

File: sktime/performance_metrics/forecasting/probabilistic/_classes.py [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+        raise TypeError("Alpha should be a float or numpy array")
+
+    def _check_y_pred_quantiles(self, y_pred):
+        assert isinstance(y_pred, pd.DataFrame)
+        assert y_pred.columns.get_level_values(0)[0] == "Quantiles"
+        pred_alphas = y_pred.columns.get_level_values(1)
+        if not all([alph in pred_alphas for alph in self.alphas]):
+            raise ValueError(f"Predictions must include the {self.alphas} quantile.")
+
+    def evaluate(self, y_true, y_pred, multioutput="uniform_average", **kwargs):
```
Comment (fkiraly) [inline sktime/performance_metrics/forecasting/probabilistic/_classes.py:129]: this feels like it shoudl be in a base class

**Review by fkiraly — APPROVED** [review]
Great, I think for a first go this is excellent.

I think we still should try refine the extension pattern though - think of a user implementing their own metric. Optimally, they would just implement `_evaluate` and `_evaluate_by_index` while inheriting from a suitable base class. Currently, that extension pattern is not possible.

I would suggest we revisit this and:
* cleaner separation between base class and implementation
* moving checks/conversions between probabilistic return types out
* writing an extension template for probabilistic metrics

I'd suggest we merge, then I take this on since I have a clear idea how these changes would look like, unless you want to look at some of the above.
